sw/airborne/modules/nav/nav_fish.c

```diff
@@ -0,0 +1,505 @@
+#endif
+
+#ifndef NAV_FISH_WALL_DISTANCE
+#define NAV_FISH_WALL_DISTANCE 10.f
+#endif
+
+#ifndef NAV_FISH_STRATEGY
+#define NAV_FISH_STRATEGY 0
+#endif
+
+#ifndef NAV_FISH_TRYATT
+#define NAV_FISH_TRYATT 0.2f
+#endif
+
+#ifndef NAV_FISH_TRLATT
+#define NAV_FISH_TRLATT 3.f
+#endif
+
+#ifndef NAV_FISH_TRYALI
+#define NAV_FISH_TRYALI 0.25f
+#endif
+
+#ifndef NAV_FISH_TRLALI
+#define NAV_FISH_TRLALI 2.f
+#endif
+
+static float distance_wall = NAV_FISH_WALL_DISTANCE;
+struct NavFishParams nav_fish_params;
+// shortname to params
+#define nfp nav_fish_params
+
+//state data
+struct NavFish {
+  float heading;      ///< heading command
+  float step_size;    ///< step size
+  float r_w;          ///< distance to wall
+  float f_w;          ///< intensity of wall effect
+  float theta_w;      ///< angle to wall
+  float f_fluct;      ///< fluctuation effect
+  float f_wall;       ///< wall effect
+  float f_ali;        ///< alignement effect
+  float f_att;        ///< attraction effect
+  float velocity;     ///< current velocity
+};
+
+struct NavFish nav_fish;
+
+/** initialization of parameters and state variables
+*/
+void nav_fish_init(void)
+{
+  nav_fish_params.max_velocity = NAV_FISH_MAXVELOCITY;
+  nav_fish_params.min_velocity = NAV_FISH_MINVELOCITY;
+  nav_fish_params.min_d2d      = NAV_FISH_MIND2D;
+  nav_fish_params.fluct        = NAV_FISH_FLUCT;
+  nav_fish_params.alpha        = NAV_FISH_ALPHA;
+  nav_fish_params.e_w1         = NAV_FISH_EW1;
+  nav_fish_params.e_w2         = NAV_FISH_EW2;
+  nav_fish_params.y_w          = NAV_FISH_YW;
+  nav_fish_params.l_w          = NAV_FISH_LW;
+  nav_fish_params.alpha_rep    = NAV_FISH_ALPHA_REP;
+  nav_fish_params.y_att        = NAV_FISH_YATT;
+  nav_fish_params.l_att        = NAV_FISH_LATT;
+  nav_fish_params.d0_att       = NAV_FISH_D0ATT;
+  nav_fish_params.y_ali        = NAV_FISH_YALI;
+  nav_fish_params.l_ali        = NAV_FISH_LALI;
+  nav_fish_params.d0_ali       = NAV_FISH_D0ALI;
+  nav_fish_params.alt          = NAV_FISH_ALT;
+  nav_fish_params.tr_y_att     = NAV_FISH_TRYATT;
+  nav_fish_params.tr_l_att     = NAV_FISH_TRLATT;
+  nav_fish_params.tr_y_ali     = NAV_FISH_TRYALI;
+  nav_fish_params.tr_l_ali     = NAV_FISH_TRLALI;
+  nav_fish_params.strategy     = NAV_FISH_STRATEGY;
+
+  nav_fish.heading = 0.f;
+  nav_fish.step_size = NAV_FISH_BODY_LENGTH;
+  nav_fish.r_w = 0.f;
+  nav_fish.f_w = 0.f;
+  nav_fish.theta_w = 0.f;
+  nav_fish.f_fluct = 0.f;
+  nav_fish.f_wall = 0.f;
+  nav_fish.f_ali = 0.f;
+  nav_fish.f_att = 0.f;
+  nav_fish.velocity = 0.5f;
+}
+
+
+static inline void send_swarm_message(void)
+{
+  DOWNLINK_SEND_SWARM_FISH(DefaultChannel, DefaultDevice,
+      &nav_fish.heading,
+      &nav_fish.step_size,
+      &nav_fish.r_w,
+      &nav_fish.f_w,
+      &nav_fish.theta_w,
+      &nav_fish.f_fluct,
+      &nav_fish.f_wall,
+      &nav_fish.f_ali,
+      &nav_fish.f_att);
+}
+
+/** sign function
+ *
+ * @param number x
+ * @return sign of x
+ */
+static float sign(float x)
```

Comment:
I do not know whether is necessary here. 

Sometimes, this sign function implementation produces chattering (jumping between -1 and 1 all the time) if we are close to zero (also I do not know whether you had any issue in your flights). I do not say you must implement it here since I do not know how easy is to implement hysteresis. Probably, it is something worth to implement in paparazzi_math xD.

Reply:
The function using this might change anyway, so I'm not sure if we will keep it in the end. But I agree that this could be integrated someday to the math lib, as well as an hysteresis function.